src/implicit_function.jl

```diff
@@ -15,21 +33,62 @@ This amounts to solving a linear system `A * J = -B`, where `A ∈ ℝᵈˣᵈ`,
 - `conditions::C`: callable of the form `(x,y,z) -> F(x,y,z)`
 - `linear_solver::L`: callable of the form `(A,b) -> u` such that `Au = b`, must be taken from Krylov.jl
 """
-struct ImplicitFunction{F,C,L}
+struct ImplicitFunction{F<:Forward,C<:Conditions,L}
     forward::F
     conditions::C
     linear_solver::L
 end
 
 """
-    ImplicitFunction(forward, conditions)
+    ImplicitFunction(forward, conditions, linear_solver, ::Val{returns_byproduct} = Val(false))
 
-Construct an `ImplicitFunction` with `Krylov.gmres` as the default linear solver.
+Construct an `ImplicitFunction` with `linear_solver` as the linear solver used in the implicit
+differentiation. By default, `ImplicitFunction(forward, conditions, linear_solver)` will assume
+that the `forward` function returns a single output and that the `conditions` function accepts
+two arguments.
+
+In some cases, it may be that a byproduct of the `forward` function can be used in the
+`conditions` function to save computation. To tell `ImplicitDifferentiation` that the
+`forward` function returns a second constant byproduct output and that the `conditions`
+function accepts that byproduct as a third argument, you can construct the implicit function
+using `ImplicitFunction(forward, conditions, linear_solver, Val(true))`.
 """
-function ImplicitFunction(forward, conditions)
+function ImplicitFunction(
+    forward, conditions, linear_solver, ::Val{returns_byproduct}=Val(false)
+) where {returns_byproduct}
+    _forward = Forward{returns_byproduct}(forward)
+    _conditions = Conditions{returns_byproduct}(conditions)
+    return ImplicitFunction(_forward, _conditions, linear_solver)
+end
+
+# JET complained without this method
```

Comment:
About what?

Reply:
Method ambiguity. I don't think there is an ambiguity.

```julia
1 ambiguities found
Ambiguity #1
ImplicitDifferentiation.ImplicitFunction(forward::F, conditions::C, linear_solver::L) where {byproduct, F<:(Forward{byproduct}), C<:(Conditions{byproduct}), L} @ ImplicitDifferentiation ~/.julia/dev/ImplicitDifferentiation/src/implicit_function.jl:37
ImplicitDifferentiation.ImplicitFunction(forward, conditions, ::Val{byproduct}) where byproduct @ ImplicitDifferentiation ~/.julia/dev/ImplicitDifferentiation/src/implicit_function.jl:84

Possible fix, define
  ImplicitDifferentiation.ImplicitFunction(::F, ::C, ::L) where {byproduct, byproduct, F<:(ImplicitDifferentiation.Forward{byproduct}), C<:(ImplicitDifferentiation.Conditions{byproduct}), L<:Val{byproduct}}

Code quality (Aqua.jl): Test Failed at /Users/mohamedtarek/.julia/packages/Aqua/2aeHf/src/ambiguities.jl:117
  Expression: success(pipeline(cmd; stdout = stdout, stderr = stderr))
```

Reply:
There is an ambiguity because `L` could be `Val{byproduct}`

Reply:
Ah yes of course